p2p/libp2p/topicProcessors_test.go

```diff
@@ -0,0 +1,138 @@
+package libp2p
+
+import (
+	"errors"
+	"testing"
+
+	"github.com/ElrondNetwork/elrond-go/core"
+	"github.com/ElrondNetwork/elrond-go/p2p"
+	"github.com/ElrondNetwork/elrond-go/p2p/mock"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestNewTopicProcessors(t *testing.T) {
+	t.Parallel()
+
+	tp := newTopicProcessors()
+
+	assert.NotNil(t, tp)
+}
+
+func TestTopicProcessorsAddShouldWork(t *testing.T) {
+	t.Parallel()
+
+	tp := newTopicProcessors()
+
+	identifier := "identifier"
+	proc := &mock.MessageProcessorStub{}
+	err := tp.addTopicProcessor(identifier, proc)
+
+	assert.Nil(t, err)
+	require.Equal(t, 1, len(tp.processors))
+	require.Equal(t, 1, len(tp.identifiers))
+	assert.True(t, proc == tp.processors[identifier]) //pointer testing
+	assert.Equal(t, identifier, tp.identifiers[0])
+}
+
+func TestTopicProcessorsDoubleAddShouldErr(t *testing.T) {
+	t.Parallel()
+
+	tp := newTopicProcessors()
+
+	identifier := "identifier"
+	_ = tp.addTopicProcessor(identifier, &mock.MessageProcessorStub{})
+	err := tp.addTopicProcessor(identifier, &mock.MessageProcessorStub{})
+
+	assert.True(t, errors.Is(err, p2p.ErrMessageProcessorAlreadyDefined))
+	require.Equal(t, 1, len(tp.processors))
+	require.Equal(t, 1, len(tp.identifiers))
+}
+
+func TestTopicProcessorsRemoveInexistentShouldErr(t *testing.T) {
+	t.Parallel()
+
+	tp := newTopicProcessors()
+
+	identifier := "identifier"
+	err := tp.removeTopicProcessor(identifier)
+
+	assert.True(t, errors.Is(err, p2p.ErrMessageProcessorDoesNotExists))
+}
+
+func TestTopicProcessorsRemoveShouldWork(t *testing.T) {
+	t.Parallel()
+
+	tp := newTopicProcessors()
+
+	identifier1 := "identifier1"
+	identifier2 := "identifier2"
+	_ = tp.addTopicProcessor(identifier1, &mock.MessageProcessorStub{})
+	_ = tp.addTopicProcessor(identifier2, &mock.MessageProcessorStub{})
+
+	require.Equal(t, 2, len(tp.processors))
+	require.Equal(t, 2, len(tp.identifiers))
+
+	err := tp.removeTopicProcessor(identifier2)
+
+	assert.Nil(t, err)
+	require.Equal(t, 1, len(tp.processors))
+	require.Equal(t, 1, len(tp.identifiers))
+
+	err = tp.removeTopicProcessor(identifier1)
+
+	assert.Nil(t, err)
+	require.Equal(t, 0, len(tp.processors))
+	require.Equal(t, 0, len(tp.identifiers))
+}
+
+func TestTopicProcessorsGetListShouldWorkAndPreserveOrder(t *testing.T) {
```

Comment:
:+1:

Reply:
if possible, please add a concurrent test for topicProcessor or directly in netMessenger where you continuously add and remove interceptors for a topic and then send a message to that topic. Code should not panic.

Reply:
Will try to assemble a test

Reply:
done